src/server/player_sao.cpp

```diff
@@ -457,20 +457,25 @@ u16 PlayerSAO::punch(v3f dir,
 
 void PlayerSAO::setHP(s32 hp, const PlayerHPChangeReason &reason)
 {
-	s32 oldhp = m_hp;
+	if (hp == (s32)m_hp)
+		return; // Nothing to do
 
-	hp = rangelim(hp, 0, m_prop.hp_max);
+	if (m_hp <= 0 && hp < (s32)m_hp)
+		return; // Cannot take more damage
 
-	if (oldhp != hp) {
-		s32 hp_change = m_env->getScriptIface()->on_player_hpchange(this, hp - oldhp, reason);
+	{
```

Comment:
Why opening a subblock?

Reply:
Mainly to isolate `hp_change` which must not be touched afterwards.